pulp_ansible/app/tasks/collections.py

```diff
@@ -205,3 +134,218 @@ def _update_highest_version(collection_version):
         collection_version.is_highest = True
         last_highest.save()
         collection_version.save()
+
+
+class AnsibleDeclarativeVersion(DeclarativeVersion):
+    """
+    Subclassed Declarative version creates a custom pipeline for RPM sync.
+    """
+
+    def pipeline_stages(self, new_version):
+        """
+        Build a list of stages feeding into the ContentUnitAssociation stage.
+
+        This defines the "architecture" of the entire sync.
+
+        Args:
+            new_version (:class:`~pulpcore.plugin.models.RepositoryVersion`): The
+                new repository version that is going to be built.
+
+        Returns:
+            list: List of :class:`~pulpcore.plugin.stages.Stage` instances
+
+        """
+        pipeline = [
+            self.first_stage,
+            QueryExistingArtifacts(),
+            ArtifactDownloader(),
+            ArtifactSaver(),
+            QueryExistingContents(),
+            AnsibleContentSaver(),
+            RemoteArtifactSaver(),
+            ResolveContentFutures(),
+        ]
+        for dupe_query_dict in self.remove_duplicates:
+            pipeline.append(RemoveDuplicates(new_version, **dupe_query_dict))
+
+        return pipeline
+
+
+class AnsibleFirstStage(Stage):
+    """
+    The first stage of a pulp_ansible sync pipeline.
+    """
+
+    def __init__(self, remote):
+        """
+        The first stage of a pulp_ansible sync pipeline.
+
+        Args:
+            remote (AnsibleRemote): The remote data to be used when syncing
+
+        """
+        super().__init__()
+        self.remote = remote
+
+        # Interpret download policy
+        self.deferred_download = self.remote.policy != Remote.IMMEDIATE
+
+    async def run(self):
+        """
+        Build and emit `DeclarativeContent` from the ansible metadata.
+        """
+        with ProgressBar(message="Parsing Collection Metadata") as pb:
+            async for metadata in self._fetch_collections():
+
+                for version in metadata["versions"]:
+
+                    response = await get_json_from_url(url=version["href"])
```

Comment:
After we merge this I want to talk on bluejeans about a design that would allow this to run in parallel instead of synchronously. For now though this is such a big step forward that I don't think we should adjust it before we merge. We can improve it later.